File: prime.js
console.log("Turing Machine for detecting prime numbers:");

const table = {
    symbols : ['0','1','X','Y','B'],
    finalstate : 6,
    transition : {
        0 : [['1','X','R'],['2','Y','R'],['6','X','R'],['6','Y','R'],['6','B','R']],
        1 : [['1','0','R'],['1','1','R'],['3','X','L'],['3','Y','L'],['3','B','L']],
        2 : [['2','0','R'],['2','1','R'],['4','X','L'],['4','Y','L'],['4','B','L']],
        3 : [['5','X','L'], [] , ['6' ,'X' ,'R'], ['6' ,'Y' ,'R'], []],
        4 : [[] ,['5' ,'Y' ,'L'] ,['6' ,'X' ,'R'], ['6' ,'Y', 'R'],[]], 
        5 : [['5','0','L'],['5','1','L'],['0','X','R'],['0','Y','R'],[]],
        6 : [[],[],[],[],[]],
    },
};

let tape = ['B','1','1','0','0','1','1','B'];
// let tape = ['B','1','1','0','0','1','0','B'];

const compute = (T) => {
    let i=1;
    let s=0;
    let red;
    let tab=[];
    let j=0;
    let map = {};
    T.symbols.forEach(sym => {
        map[sym]=j++;
    });
    while(s!=T.finalstate){
        red = tape[i];
        tab = T.transition[s][map[red]];
        if(tab.length==0){
            console.log("String does not match grammer");
            return;
        }
        s = tab[0];
        tape[i] = tab[1];
        i+= tab[2]=='R'?1:-1;
        tape.forEach(element => {
            process.stdout.write(element+" ");
        });
        console.log("\n");
        for(let it=0;it<tape.length;it++){
            let out = (it==i)?"^ ":"  ";
            process.stdout.write(out);
        }
        console.log("\n");
    }
    console.log("String matches the Grammar");
};

compute(table,tape);

File: prime.cpp
#include<iostream>
#include<stdlib.h>
#include<bits/stdc++.h>
#include<string>
const int LIM = 100;
using namespace std;

class table{
    public:
    bool empty;
    int state;
    char output;
    bool dir;
    table(){
    }
    table(int s,char c,bool b,bool emp=false){
        state = s;
        output = c;
        dir = b;
        empty = emp;
    }
};

class turing{
    public:
    vector<vector<table>> t;
    unordered_map<char,int> m;
    vector<char> tape; 
    int numofstates;
    int symbols;
    int finalstate;
    int count;
    turing(){
        count = 0;
    }
    void inputlanguage();
    void print(){
        for(auto & x: m){
            cout<<x.first<<" ";
        }
        cout<<endl;
        for(auto &x : t){
            for( auto &i : x){
                cout<<i.state<<i.output<<i.dir<<" ";
            }
            cout<<endl;
        }
    }
    void gettape(){
        int num;
        cin>>num;
        tape.push_back('B');
        for(int i=0;i<num;i++){
            char k;
            cin>>k;
            tape.push_back(k);
        }
        tape.push_back('B');
    }
    void printtape(){
        for(auto &x:tape){
            cout<<x<<" ";
        }
        cout<<endl;
    }
    void compute(int N=0){
        int i = 1;
        int s = 0;
        char red;
        table match;
        while(s!=finalstate){
            if(i==-1){
                tape.insert(tape.begin(),'B');
                i++;
            }
            if(i==tape.size()){
                tape.push_back('B');
            }
            red = tape[i];
            match = t[s][m[red]];
            if(match.empty) {
                //cout<<N<<" is NOT Prime."<<endl;
                cout<<"is not valid"<<endl;
                return;
            }
            s = match.state;
            tape[i] = match.output;
            i += (match.dir)?1:-1;
            printtape();
            for(int it=0;it<tape.size();it++){
                if(it==i) cout<<"^ ";
                else cout<<"  ";
            }
            cout<<endl;
        }
        //cout<<N<<" is Prime."<<endl;
        //cout<<N<<endl;
        cout<<"is valid."<<endl;
        count++;
    }
    void check_primes(){
        for(int i=0;i<LIM;i++){
            tape.clear();
            tape.push_back('B');
            for(int it = 0;it<i;it++){
                tape.push_back('0');
            }
            tape.push_back('1');
            for(int it = 0;it<i;it++){
                tape.push_back('0');
            }
            tape.push_back('B');
            compute(i);
        }
    }
};

void turing::inputlanguage(){
    cin>>numofstates>>symbols;
    for(int i=0;i<symbols;i++){
        char ch;
        cin>>ch;
        m[ch] = i;
    }
    for(int i=0;i<numofstates;i++){
        int xe;
        int tmp;
        cin>>xe;
        t.push_back(vector<table>());
        for(int en=0;en<symbols;en++){
            // char check;
            string check;
            char ch;
            cin>>check;
            if(check == "-"){
                table temp(0,'-',0,1);
                t[i].push_back(temp);
                cin>>check;
                cin>>check;
                continue;
            }
            int state = stoi(check);
            cin>>ch;
            table temp(state,ch,0);
            cin>>ch;
            if(ch=='R'){
                temp.dir = true;
            }
            else temp.dir = false;
            t[i].push_back(temp);
        }
    }
    cin>>finalstate;
}

int main(){
    turing t;
    t.inputlanguage();
    t.gettape();
    t.compute();
    // t.check_primes();
    // cout<<"Number of primes are: "<<t.count<<endl;
}